networkx/algorithms/isomorphism/tests/test_tree_isomorphism.py

```diff
@@ -154,127 +114,66 @@ def test_hardcoded():
     t2.add_edges_from(edges_2)
     root2 = "n"
 
-    isomorphism = sorted(rooted_tree_isomorphism(t1, root1, t2, root2))
+    isomorphism = sorted(nx.isomorphism.rooted_tree_isomorphism(t1, root1, t2, root2))
 
     # is correct by hand
     assert isomorphism in (isomorphism1, isomorphism2)
 
     # check algorithmically
-    assert check_isomorphism(t1, t2, isomorphism)
-
-
-# randomly swap a tuple (a,b)
-def random_swap(t):
-    (a, b) = t
-    if random.randint(0, 1) == 1:
-        return (a, b)
-    else:
-        return (b, a)
-
-
-# given a tree t1, create a new tree t2
-# that is isomorphic to t1, with a known isomorphism
-# and test that our algorithm found the right one
-def positive_single_tree(t1):
-    assert nx.is_tree(t1)
-
-    nodes1 = list(t1.nodes())
-    # get a random permutation of this
-    nodes2 = nodes1.copy()
-    random.shuffle(nodes2)
-
-    # this is one isomorphism, however they may be multiple
-    # so we don't necessarily get this one back
-    someisomorphism = list(zip(nodes1, nodes2))
-
-    # map from old to new
-    map1to2 = dict(someisomorphism)
-
-    # get the edges with the transformed names
-    edges2 = [random_swap((map1to2[u], map1to2[v])) for (u, v) in t1.edges()]
-    # randomly permute, to ensure we're not relying on edge order somehow
-    random.shuffle(edges2)
-
-    # so t2 is isomorphic to t1
-    t2 = nx.Graph()
-    t2.add_edges_from(edges2)
-
-    # lets call our code to see if t1 and t2 are isomorphic
-    isomorphism = tree_isomorphism(t1, t2)
-
-    # make sure we got a correct solution
-    # although not necessarily someisomorphism
-    assert len(isomorphism) > 0
-    assert check_isomorphism(t1, t2, isomorphism)
-
-
-# run positive_single_tree over all the
-# non-isomorphic trees for k from 4 to maxk
-# k = 4 is the first level that has more than 1 non-isomorphic tree
-# k = 13 takes about 2.86 seconds to run on my laptop
-# larger values run slow down significantly
-# as the number of trees grows rapidly
-def test_positive(maxk=14):
-    for k in range(2, maxk + 1):
-        start_time = time.time()
-        trial = 0
-        for t in nx.nonisomorphic_trees(k):
-            positive_single_tree(t)
-            trial += 1
-
-
-# test the trivial case of a single node in each tree
-# note that nonisomorphic_trees doesn't work for k = 1
-def test_trivial():
-    # back to an undirected graph
+    assert _check_isomorphism(t1, t2, isomorphism)
+
+
+# NOTE: number of nonisomorphic_trees grows very rapidly - do not increase n
+# further without marking "slow"
+@pytest.mark.parametrize("n", range(2, 15))
+def test_tree_isomorphic_all_non_isomorphic_trees_relabeled(n):
+    """Tests every non-isomorphic tree with `n` nodes is isomorphic with a
+    copy of itself with an arbitrary node-remapping."""
+    for tree in nx.nonisomorphic_trees(n):
+        nodes = list(tree)
+        random.shuffle(shuffled := nodes.copy())
```

Comment:
I added this because it occurred to me that you could use the walrus operator to save lines when using in-place operations. That said, I don't necessarily like this pattern (I'd vastly prefer to use a non-in-place shuffle (like `np.permuted`), but AFAIK that doesn't exist in the `random` module.

The walrus stuff (see also new_edges below) is self-contained in 5cb7dcf for easy reversion :)

Reply:
I think the way to return a new list with "shuffled" values is `new_nodes = random.sample(nodes, len(nodes))`.  The word "shuffle" is so good... They should have made a function `shuffled` to go with it.  But the word "sample" is reasonable too I guess -- and extends to any size samples without replacement.

I'm fine with the walrus approach too. But wanted to point out the `sample` function within `random`.